Initial documentation for SEA (#413)
microsoft/mu_feature_mm_supv: SeaPkg/Docs/PlatformIntegration/PlatformIntegrationSteps.md

to the required code changes to integrate all of the pieces.

## High-Level Considerations

1. [MM Supervisor Changes](#mm-supervisor-changes) - The supervisor is accountable for loading SEA core into MSEG region
and protect the resources before SEA is loaded.

1. [Executed Supervisor Validation](#executed-supervisor-validation) - The SEA core is responsible for inspecting the MM
supervisor environment and validating the MM supervisor code and data. This is a critical step in the DRTM process.

1. [Platform Data Requirements](#platform-data-requirements) - The MM Supervisor requires a new set of industry standard
defined data structures in addition to supervisor-specific data structures to be produced by the platform.

1. [SEA Code Integration](#sea-code-integration) - How to best integrate the `SEA core` collateral into a platform firmware.

## MM Supervisor Changes

Begin by reading the [MM supervisor documentation](../../../MmSupervisorPkg/Docs/MmSupervisor_Feature.md) to gain a basic
understanding of Standalone MM and MM supervisor.

In the SEA approach, the SEA is responsible for validating the entire flow of the MMI, which starts from the MMI entrypoint,

> jumping to the MM supervisor, and ended with return. This will require the MM entrypoint code to be decoupled from the MM

Reviewer: ```suggestion
jumps into the MM supervisor and ends by returning. This will require the MM entry point code to be decoupled from the MM
```

Reviewer: Included in latest push

supervisor, and the new MMI entrypoint code is auditable with respect to the jump point into the MM supervisor.

Hence, a new [MMI entrypoint](../../MmiEntrySea/MmiEntrySea.inf) is created to allow the MM supervisor to load using MM
relocation phase. It is also crafted so that the jump point into the MM supervisor is located at the end of the MMI entrypoint
code. This allows the SEA to validate the entire flow of the MMI.

Thus, a new [SmmCpuFeaturesLib](../../Library/SmmCpuFeaturesLib/StandaloneMmCpuFeaturesLibStm.inf)
is created to provide the necessary CPU features for the MM Supervisor to load the SEA core.

Specifically, the library, linked to the MM supervisor, will locate the resources needed, load the SEA core into MSEG, load
the MMI entrypoint code blob to MM_BASE for each core, fix up the jump points based on MM supervisor function pointers, and
finally protected the loaded images.

## Executed Supervisor Validation

> As part of a DRTM effort, the SEA core is respoinsble for verifying the entire environment of the MM supervisor. As DRTM

Reviewer 2: ```suggestion
As part of a DRTM effort, the SEA core is responsible for verifying the entire environment of the MM supervisor. As DRTM

```

Reviewer: Fixed in latest push

"Verify the MM Core code through a hash."

__VerifyAndHashImage Flow__:

The VerifyAndHashImage function follows this flow:

1. Verify that the image is in MMRAM.
1. Copy the image over to MSEG and retrieve the image information.
1. Create a buffer to copy the image into and then validate the differences found based on the AUX file.
    - Everything labeled as a Rule is validated based on the type of rule, and then the contents of the original binary
    are copied over to maintain coherency for the hash validation.
1. Revert the loaded image and place it into a newly created buffer.
1. Hash the reverted image and compare it to the original hash of the supervisor binary.
1. Reversion Process:

The behavior of the reversion code is somewhat complex and relies on reported image sections. This leads to different
behavior between the DEBUG and RELEASE builds.

__Reversion Flow__:

1. Validate section alignment.
1. Read the PE/COFF header into memory and read the first section, which provides the number of sections.
1. Loop through each section of the image, copying over the sections if they have a SizeOfRawData > 0.

> 1. The behavior differs between the DEBUG and RELEASE builds, as highlighted below.

Reviewer 2: This should be a subpoint of the line above it ("Loop through each section...")

Reviewer 2: I see when actually viewing the document it looks correct so I'm resolving this comment.

Reviewer 2: Maybe reword it slightly as well.
```suggestion
     - This behavior differs betweent he DEBUG and RELEASE builds.

```

Reviewer: Fixed in latest push

As defined in this [header file](../../Include/SeaAuxiliary.h), the SEA core will use the following rules to validate the
MM memory regions:

1. `IMAGE_VALIDATION_ENTRY_TYPE_NONE`: Ignore the region corresponding to this entry.
1. `IMAGE_VALIDATION_ENTRY_TYPE_MSEG`: The region is MSEG, SEA core will validate the region to be MSEG.
1. `IMAGE_VALIDATION_ENTRY_TYPE_CONTENT`: The content inside this region matches exactly the content inside the rule.
1. `IMAGE_VALIDATION_ENTRY_TYPE_MEM_ATTR`: The memory attribute of the region matches the rule.
1. `IMAGE_VALIDATION_ENTRY_TYPE_SELF_REF`: The region is self-referenced, SEA core will validate the region to be pointing
to another point inside the MM supervisor.
1. `IMAGE_VALIDATION_ENTRY_TYPE_POINTER`: The entry is a pointer to and the pointer needs to be not null, pointing to non
MSEG, or pointing to outside of TSEG, depending on the rule.

The syntax of the rules is as described in this [documentation](../../Tools/GenSeaArtifacts/gen_aux/readme.md).

## SEA Code Integration

A general description of SEA and MM supervisor code integration is depicted below:

![SEA and MM Supervisor Assert Layout](Images/asset_layout.png)

> 1. Ensure all submodules for the platform are based on the latest Project Mu version (e.g. "202102")

Reviewer 2: Probably should use the most up-to-date mu verison.

Reviewer 2: Especially because SEA wouldn't work with 202102.

Reviewer: Updated in latest push
